docs/ABI.rst

```diff
@@ -909,92 +936,120 @@ constraints on the extension are mangled in its generic signature.
+  known-nominal-type ::= 'Sd'                // Swift.Float64
+  known-nominal-type ::= 'Sf'                // Swift.Float32
+  known-nominal-type ::= 'Si'                // Swift.Int
+  known-nominal-type ::= 'SV'                // Swift.UnsafeRawPointer
+  known-nominal-type ::= 'Sv'                // Swift.UnsafeMutableRawPointer
+  known-nominal-type ::= 'SP'                // Swift.UnsafePointer
+  known-nominal-type ::= 'Sp'                // Swift.UnsafeMutablePointer
+  known-nominal-type ::= 'SQ'                // Swift.ImplicitlyUnwrappedOptional
+  known-nominal-type ::= 'Sq'                // Swift.Optional
+  known-nominal-type ::= 'SR'                // Swift.UnsafeBufferPointer
+  known-nominal-type ::= 'Sr'                // Swift.UnsafeMutableBufferPointer
+  known-nominal-type ::= 'SS'                // Swift.String
+  known-nominal-type ::= 'Su'                // Swift.UInt
+
+  protocol ::= context decl-name
+
   type ::= 'Bb'                              // Builtin.BridgeObject
   type ::= 'BB'                              // Builtin.UnsafeValueBuffer
-  type ::= 'Bf' natural '_'                  // Builtin.Float<n>
-  type ::= 'Bi' natural '_'                  // Builtin.Int<n>
+  type ::= 'Bf' NATURAL '_'                  // Builtin.Float<n>
+  type ::= 'Bi' NATURAL '_'                  // Builtin.Int<n>
   type ::= 'BO'                              // Builtin.UnknownObject
   type ::= 'Bo'                              // Builtin.NativeObject
   type ::= 'Bp'                              // Builtin.RawPointer
-  type ::= 'Bv' natural type                 // Builtin.Vec<n>x<type>
+  type ::= type 'Bv' NATURAL                 // Builtin.Vec<n>x<type>
   type ::= 'Bw'                              // Builtin.Word
-  type ::= nominal-type
-  type ::= associated-type
-  type ::= 'a' context identifier            // Type alias (DWARF only)
-  type ::= 'b' type type                     // objc block function type
-  type ::= 'c' type type                     // C function pointer type
-  type ::= 'F' throws-annotation? type type  // function type
-  type ::= 'f' throws-annotation? type type  // uncurried function type  
-  type ::= 'G' type <type>+ '_'              // generic type application
-  type ::= 'K' type type                     // @auto_closure function type
-  type ::= 'M' type                          // metatype without representation
-  type ::= 'XM' metatype-repr type           // metatype with representation
-  type ::= 'P' protocol-list '_'             // protocol type
-  type ::= 'PM' type                         // existential metatype without representation
-  type ::= 'XPM' metatype-repr type          // existential metatype with representation
+  type ::= context identifier 'a'            // Type alias (DWARF only)
+  type ::= type type 'c' THROWS-ANNOTATION?  // function type
+  type ::= type type 'cu' THROWS-ANNOTATION?  // uncurried function type  
```

Comment:
Function types should have a variable number of input arguments, instead of a single tuple argument. "Variadic" and "inout" should be argument modifiers on function types instead of separate type encodings.